runtime/drivers/duckdb/olap.go

```diff
@@ -565,6 +584,48 @@ func (c *connection) updateVersion(name, version string) error {
 	return err
 }
 
+func (c *connection) execWithLimits(parentCtx context.Context, stmt *drivers.Statement) error {
+	storageLimit := c.config.StorageLimitBytes
+	if storageLimit <= 0 { // no limit
+		return c.Exec(parentCtx, stmt)
+	}
+
+	// check current size
+	currentSize, _ := c.EstimateSize()
+	storageLimit -= currentSize
+	// current size already exceeds limit
+	if storageLimit <= 0 {
+		return drivers.ErrStorageLimitExceeded
+	}
+
+	ctx, cancel := context.WithCancel(parentCtx)
+	limitExceeded := atomic.Bool{}
+	// Start background goroutine to check size is not exceeded during query execution
+	go func() {
+		ticker := time.NewTicker(5 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				if size, ok := c.EstimateSize(); ok && size > storageLimit {
+					limitExceeded.Store(true)
+					cancel()
+					return
+				}
+			}
+		}
+	}()
+
+	err := c.Exec(ctx, stmt)
+	cancel()
```

Comment:
nit: move to a `defer cancel()` right after call to `WithCancel`. It's safe to call multiple times and ensures it will always be called even if someone adds other conditions in the future.